mmeds/tests/unit/test_demultiplex.py

```diff
@@ -0,0 +1,118 @@
+        cls.strip_dir.mkdir(exist_ok=True)
+        cls.log_dir.mkdir(exist_ok=True)
+        cls.out = Path(cls.pheniqs_dir) / 'pheniqs_config_test.json'
+        cls.log = Path(cls.log_dir) / 'pheniqs_report.txt'
+
+    def test_pheniqs(cls):
+        """ Test pheniqs demultiplexing """
+        create_pheniqs_config(cls)
+        new_env = setup_environment('pheniqs/2.1.0')
+
+        gunzip_forward_barcodes = ['gunzip', f'{cls.for_barcodes}.gz']
+        gunzip_reverse_barcodes = ['gunzip', f'{cls.rev_barcodes}.gz']
+        pheniqs_demultiplex = ['pheniqs', 'mux', '--config', f'{cls.out}']
+
+        try:
+            run(pheniqs_demultiplex, capture_output=True, env=new_env, check=True)
+            run(gunzip_forward_barcodes, capture_output=True, check=True)
+            run(gunzip_reverse_barcodes, capture_output=True, check=True)
+
+        except CalledProcessError as e:
+            Logger.debug(e)
+            print(e.output)
+
+        # validate one of the demultiplexed files
+        validate_demultiplex(f'{cls.out_dir}/{cls.target_file}', cls.for_barcodes,
+                             cls.rev_barcodes, cls.mapping, cls.log_dir, True)
+
+        # check values in the validation log file
+        validate_log = cls.log_dir / f'{cls.target_file}_report.log'
+        log_df = pd.read_csv(validate_log, sep=':', skiprows=[0], names=['stat', 'value'])
+        log_df.set_index('stat', inplace=True)
+        log_df = log_df.T
+
+        cls.assertEqual(log_df['Percent duplicate labels']['value'], 0)
+        cls.assertEqual(log_df['Percent QIIME-incompatible fasta labels']['value'], 0)
+
+        # If it's 1, no SampleIDs were found
+        # This should be a value between 0 and 1, representing the percentage of exact match barcodes
+        cls.assertNotEqual(log_df['Percent of labels that fail to map to SampleIDs']['value'], 1)
+        cls.assertEqual(log_df['Percent of sequences with invalid characters']['value'], 0)
+        cls.assertEqual(log_df['Percent of sequences with barcodes detected']['value'], 0)
+        cls.assertEqual(
+            log_df['Percent of sequences with barcodes detected at the beginning of the sequence']['value'], 0)
+
+        cls.assertEqual(log_df['Percent of sequences with primers detected']['value'], 0)
+        cls.assertTrue('All SampleIDs found in sequence labels.' in log_df.columns)
+
+        gzip1 = ['gzip', f'{cls.for_barcodes}']
+        gzip2 = ['gzip', f'{cls.rev_barcodes}']
+        try:
+            # pass
+            run(gzip1, capture_output=True, check=True)
+            run(gzip2, capture_output=True, check=True)
+        except CalledProcessError as e:
+            Logger.debug(e)
+            print(e.output)
+
+    @classmethod
+    def tearDownClass(cls):
+        """ Set up tests """
+        # cleanup demultiplexed files.
+        remove_test_dir = ['rm', '-rf', f'{cls.out_dir}']
+        try:
+            run(remove_test_dir, capture_output=True, check=True)
+
```

Comment:
this is formatted a little weird, just make the newlines look a little nicer. but also, what would it mean if this error was caught? should the test fail if it does?

Reply:
hmm I think the logic around doing subprocesses in try-excepts is that otherwise it's hard to get info if they do fail. Don't think the test should fail if this fails, also what newlines?